File: V2rayClient/Services/RoutingRuleManager.cs
using System.IO;
using V2rayClient.Models;
using V2rayClient.Models.SingBox;

namespace V2rayClient.Services;

/// <summary>
/// Service for managing sing-box routing rules
/// </summary>
public class RoutingRuleManager : IRoutingRuleManager
{
    /// <summary>
    /// Event raised when routing rules are updated and require restart
    /// </summary>
    public event EventHandler<RoutingRulesChangedEventArgs>? RoutingRulesChanged;

    /// <summary>
    /// Notify that routing rules have been updated
    /// </summary>
    /// <param name="changeType">Type of change that occurred</param>
    public void NotifyRulesChanged(RuleChangeType changeType)
    {
        RoutingRulesChanged?.Invoke(this, new RoutingRulesChangedEventArgs
        {
            ChangeType = changeType,
            Timestamp = DateTime.Now
        });
    }

    /// <summary>
    /// Generate sing-box format routing configuration based on proxy mode and custom rules.
    /// This is the primary method for generating routing rules for sing-box core.
    /// </summary>
    /// <param name="mode">Proxy mode (Global, Smart, Direct)</param>
    /// <param name="customRules">List of custom domain rules</param>
    /// <returns>RouteConfig for sing-box configuration</returns>
    public RouteConfig GenerateSingBoxRouting(ProxyMode mode, List<DomainRule> customRules)
    {
        var routeConfig = new RouteConfig
        {
            Rules = new List<RouteRule>(),
            DefaultDomainResolver = "dns-local"
        };

        // Priority 1: Custom domain rules (highest priority)
        routeConfig.Rules.AddRange(GenerateSingBoxCustomRules(customRules));

        // Priority 2: Mode-specific rules
        switch (mode)
        {
            case ProxyMode.Global:
                routeConfig.Final = "proxy";
                break;
            case ProxyMode.Smart:
                routeConfig.RuleSet = GenerateSingBoxRuleSets();
                routeConfig.Rules.AddRange(GenerateSingBoxSmartModeRules());
                routeConfig.Final = "proxy";
                break;
            case ProxyMode.Direct:
                routeConfig.Final = "direct";
                break;
        }

        return routeConfig;
    }

    /// <summary>
    /// Generate sing-box routing rules from custom domain rules
    /// </summary>
    private List<RouteRule> GenerateSingBoxCustomRules(List<DomainRule> customRules)
    {
        var rules = new List<RouteRule>();

        if (customRules == null || customRules.Count == 0)
        {
            return rules;
        }

        // Process each custom rule individually to maintain order
        foreach (var customRule in customRules.Where(r => r.Enabled))
        {
            var (exactDomains, suffixDomains) = SplitDomainsBySingBoxFormat(customRule.Domains);
            
            var (action, outbound) = customRule.Strategy switch
            {
                RuleStrategy.Proxy => ("route", "proxy"),
                RuleStrategy.Direct => ("route", "direct"),
                RuleStrategy.Block => ("reject", null),
                _ => ("route", "direct")
            };
            
            var rule = new RouteRule { Action = action };
            if (outbound != null) rule.Outbound = outbound;
            if (exactDomains.Count > 0) rule.Domain = exactDomains;
            if (suffixDomains.Count > 0) rule.DomainSuffix = suffixDomains;
            
            rules.Add(rule);
        }

        return rules;
    }

    /// <summary>
    /// Generate sing-box rule sets for Smart mode
    /// </summary>
    private List<RuleSet> GenerateSingBoxRuleSets()
    {
        var geositeDir = App.ResourceManager?.GeositeDir 
            ?? Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Resources", "geosite");
        
        return new List<RuleSet>
        {
            new RuleSet
            {
                Tag = "geosite-cn",
                Type = "local",
                Format = "binary",
                Path = Path.Combine(geositeDir, "geosite-cn.srs")
            },
            new RuleSet
            {
                Tag = "geosite-geolocation-!cn",
                Type = "local",
                Format = "binary",
                Path = Path.Combine(geositeDir, "geosite-geolocation-!cn.srs")
            },
            new RuleSet
            {
                Tag = "geoip-cn",
                Type = "local",
                Format = "binary",
                Path = Path.Combine(geositeDir, "geoip-cn.srs")
            }
        };
    }

    /// <summary>
    /// Generate sing-box routing rules for Smart mode (China direct, others proxy)
    /// </summary>
    private List<RouteRule> GenerateSingBoxSmartModeRules()
    {
        return new List<RouteRule>
        {
            // DNS hijacking
            new RouteRule
            {
                Protocol = "dns",
                Action = "hijack-dns"
            },
            // Block QUIC to prevent UDP-based protocols
            new RouteRule
            {
                Protocol = "quic",
                Action = "reject"
            },
            // Foreign domains go proxy
            new RouteRule
            {
                RuleSet = "geosite-geolocation-!cn",
                Action = "route",
                Outbound = "proxy"
            },
            // Chinese domains go direct
            new RouteRule
            {
                RuleSet = "geosite-cn",
                Action = "route",
                Outbound = "direct"
            },
            // Chinese IPs go direct
            new RouteRule
            {
                RuleSet = "geoip-cn",
                Action = "route",
                Outbound = "direct"
            }
        };
    }

    /// <summary>
    /// Split domains into exact match and suffix match for sing-box format
    /// </summary>
    /// <param name="domains">List of domain patterns</param>
    /// <returns>Tuple of (exact domains, suffix domains)</returns>
    private (List<string> exactDomains, List<string> suffixDomains) SplitDomainsBySingBoxFormat(List<string> domains)
    {
        var exactDomains = new List<string>();
        var suffixDomains = new List<string>();

        foreach (var domain in domains)
        {
            if (string.IsNullOrWhiteSpace(domain))
            {
                continue;
            }

            var trimmedDomain = domain.Trim().ToLowerInvariant();

            // If domain starts with *., it's a suffix match
            // *.example.com -> example.com (matches example.com and all subdomains)
            if (trimmedDomain.StartsWith("*."))
            {
                suffixDomains.Add(trimmedDomain.Substring(2));
            }
            else
            {
                // Exact domain match
                exactDomains.Add(trimmedDomain);
            }
        }

        return (exactDomains, suffixDomains);
    }

    /// <summary>
    /// [Obsolete] Generate v2ray-core format routing rules.
    /// This method is deprecated and kept only for backward compatibility.
    /// Use GenerateSingBoxRouting instead for all new code.
    /// </summary>
    /// <param name="mode">Proxy mode (Global, Smart, Direct)</param>
    /// <param name="customRules">List of custom domain rules</param>
    /// <returns>Empty list (v2ray-core support has been removed)</returns>
    [Obsolete("This method is deprecated. Use GenerateSingBoxRouting instead. V2ray-core support has been removed in favor of sing-box.")]
    public List<object> GenerateRoutingRules(ProxyMode mode, List<DomainRule> customRules)
    {
        // V2ray-core support has been removed. This method returns an empty list for backward compatibility.
        // All routing should now use GenerateSingBoxRouting which returns RouteConfig for sing-box.
        return new List<object>();
    }
}

File: src/main/services/RoutingRuleManager.ts
/**
 * 路由规则管理服务
 * 负责生成 sing-box 路由规则配置
 */

import type { UserConfig, DomainRule } from '../../shared/types';

/**
 * sing-box 路由规则接口
 */
export interface SingBoxRouteRule {
  // 域名匹配
  domain?: string[];
  domain_suffix?: string[];
  domain_keyword?: string[];
  domain_regex?: string[];

  // GeoIP/GeoSite 匹配
  geosite?: string[];
  geoip?: string[];

  // IP 匹配
  ip_cidr?: string[];

  // 端口匹配
  port?: number[];
  port_range?: string[];

  // 协议匹配
  protocol?: string[];

  // 出站标签
  outbound: string;
}

/**
 * sing-box 路由配置接口
 */
export interface SingBoxRouteConfig {
  rules: SingBoxRouteRule[];
  rule_set?: Array<{
    tag: string;
    type: string;
    format: string;
    path?: string;
    url?: string;
  }>;
  final?: string;
  auto_detect_interface?: boolean;
}

export interface IRoutingRuleManager {
  /**
   * 生成路由配置
   */
  generateRouteConfig(config: UserConfig): SingBoxRouteConfig;

  /**
   * 生成全局代理模式规则
   */
  generateGlobalProxyRules(): SingBoxRouteRule[];

  /**
   * 生成智能分流模式规则
   */
  generateSmartRoutingRules(): SingBoxRouteRule[];

  /**
   * 生成直连模式规则
   */
  generateDirectRules(): SingBoxRouteRule[];

  /**
   * 集成自定义域名规则
   */
  integrateCustomRules(customRules: DomainRule[]): SingBoxRouteRule[];
}

export class RoutingRuleManager implements IRoutingRuleManager {
  /**
   * 生成路由配置
   */
  generateRouteConfig(config: UserConfig): SingBoxRouteConfig {
    const rules: SingBoxRouteRule[] = [];

    // 1. 首先添加自定义规则（优先级最高）
    const customRules = this.integrateCustomRules(config.customRules);
    rules.push(...customRules);

    // 2. 根据代理模式添加相应规则
    let modeRules: SingBoxRouteRule[] = [];
    switch (config.proxyMode) {
      case 'global':
        modeRules = this.generateGlobalProxyRules();
        break;
      case 'smart':
        modeRules = this.generateSmartRoutingRules();
        break;
      case 'direct':
        modeRules = this.generateDirectRules();
        break;
    }
    rules.push(...modeRules);

    // 3. 构建完整的路由配置
    const routeConfig: SingBoxRouteConfig = {
      rules,
      auto_detect_interface: true,
    };

    // 4. 设置默认出站
    if (config.proxyMode === 'direct') {
      routeConfig.final = 'direct';
    } else {
      routeConfig.final = 'proxy';
    }

    return routeConfig;
  }

  /**
   * 生成全局代理模式规则
   * 所有流量都通过代理，除了本地地址
   */
  generateGlobalProxyRules(): SingBoxRouteRule[] {
    return [
      // 本地地址直连
      {
        ip_cidr: [
          '127.0.0.0/8',
          '10.0.0.0/8',
          '172.16.0.0/12',
          '192.168.0.0/16',
          '169.254.0.0/16',
          'fc00::/7',
          'fe80::/10',
          '::1/128',
        ],
        outbound: 'direct',
      },
      // 本地域名直连
      {
        domain_suffix: ['localhost', 'local'],
        outbound: 'direct',
      },
    ];
  }

  /**
   * 生成智能分流模式规则
   * 使用 GeoIP 和 GeoSite 进行智能分流
   */
  generateSmartRoutingRules(): SingBoxRouteRule[] {
    return [
      // 本地地址直连
      {
        ip_cidr: [
          '127.0.0.0/8',
          '10.0.0.0/8',
          '172.16.0.0/12',
          '192.168.0.0/16',
          '169.254.0.0/16',
          'fc00::/7',
          'fe80::/10',
          '::1/128',
        ],
        outbound: 'direct',
      },
      // 本地域名直连
      {
        domain_suffix: ['localhost', 'local'],
        outbound: 'direct',
      },
      // 中国大陆域名直连
      {
        geosite: ['cn'],
        outbound: 'direct',
      },
      // 中国大陆 IP 直连
      {
        geoip: ['cn'],
        outbound: 'direct',
      },
      // 国外域名走代理
      {
        geosite: ['geolocation-!cn'],
        outbound: 'proxy',
      },
    ];
  }

  /**
   * 生成直连模式规则
   * 所有流量都直连
   */
  generateDirectRules(): SingBoxRouteRule[] {
    // 直连模式不需要特殊规则，所有流量都走 direct
    return [];
  }

  /**
   * 集成自定义域名规则
   * 将用户自定义的域名规则转换为 sing-box 规则格式
   * 所有域名统一使用 domain_suffix 匹配，即匹配该域名及其所有子域名
   */
  integrateCustomRules(customRules: DomainRule[]): SingBoxRouteRule[] {
    // 按 action 分组规则
    const rulesByAction = new Map<string, string[]>();

    for (const rule of customRules) {
      // 只处理启用的规则
      if (!rule.enabled) {
        continue;
      }

      const action = rule.action;
      if (!rulesByAction.has(action)) {
        rulesByAction.set(action, []);
      }

      // 支持 domains 数组，统一去掉 *. 前缀
      for (const domain of rule.domains) {
        const cleanDomain = domain.startsWith('*.') ? domain.slice(2) : domain;
        rulesByAction.get(action)!.push(cleanDomain);
      }
    }

    // 转换为 sing-box 规则
    const singboxRules: SingBoxRouteRule[] = [];

    for (const [action, domains] of rulesByAction.entries()) {
      if (domains.length === 0) {
        continue;
      }

      // 确定出站标签
      let outbound: string;
      if (action === 'proxy') {
        outbound = 'proxy';
      } else if (action === 'direct') {
        outbound = 'direct';
      } else if (action === 'block') {
        outbound = 'block';
      } else {
        // 未知 action，跳过
        continue;
      }

      // 统一使用 domain_suffix，匹配域名及其所有子域名
      const rule: SingBoxRouteRule = {
        outbound,
        domain_suffix: domains,
      };

      singboxRules.push(rule);
    }

    return singboxRules;
  }
}
